staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -407,10 +407,51 @@ func (qs *queueSet) lockAndSyncTime() {
 // lock and before modifying the state of any queue.
 func (qs *queueSet) syncTimeLocked() {
 	realNow := qs.clock.Now()
-	timeSinceLast := realNow.Sub(qs.lastRealTime).Seconds()
+	timeSinceLast := realNow.Sub(qs.lastRealTime)
 	qs.lastRealTime = realNow
-	qs.virtualTime += timeSinceLast * qs.getVirtualTimeRatioLocked()
-	metrics.SetCurrentR(qs.qCfg.Name, qs.virtualTime)
+	prevR := qs.currentR
+	incrR := SeatsTimesDuration(qs.getVirtualTimeRatioLocked(), timeSinceLast)
+	if incrR > RDecrement/2 {
+		klog.ErrorS(errors.New("unusably large R increment"), "Can not do it", "QS", qs.qCfg.Name, "when", realNow.Format(nsTimeFmt), "incrR", incrR)
+	}
+	qs.currentR = prevR + incrR
+	switch {
+	case prevR > qs.currentR:
+		klog.ErrorS(errors.New("progress meter overflow"), "Wrap", "QS", qs.qCfg.Name, "when", realNow.Format(nsTimeFmt), "prevR", prevR, "currentR", qs.currentR)
+	case qs.currentR >= HighR:
+		qs.advanceEpoch()
+	}
+	metrics.SetCurrentR(qs.qCfg.Name, qs.currentR.ToFloat())
+}
+
+// advanceEpoch subtracts RDecrement from the global progress meter R
```

Comment:
Can you please address https://github.com/kubernetes/kubernetes/pull/105412#discussion_r721964269
in this PR?

Reply:
I expanded the test and added an explanation of why it is the way it is and why that should work.